cpp/src/arrow/util/async_generator.h

```diff
@@ -280,6 +280,150 @@ AsyncGenerator<V> MakeMappedGenerator(AsyncGenerator<T> source_generator,
   return MappingGenerator<T, V>(std::move(source_generator), std::move(map));
 }
 
+/// \see MakeSequencingGenerator
+template <typename T, typename Comp, typename IsNext>
+class SequencingGenerator {
+ public:
+  SequencingGenerator(AsyncGenerator<T> source, Comp compare, IsNext is_next,
+                      T initial_value)
+      : state_(std::make_shared<State>(std::move(source), std::move(compare),
+                                       std::move(is_next), std::move(initial_value))) {}
+
+  Future<T> operator()() {
+    util::optional<Result<T>> result;
+    {
+      auto guard = state_->mutex.Lock();
+      if (!state_->queue.empty()) {
+        if (!state_->queue.top().ok() ||
+            state_->is_next(state_->previous_value, *state_->queue.top())) {
+          result = std::move(state_->queue.top());
+          if (result->ok()) {
+            state_->previous_value = **result;
+          }
+          state_->queue.pop();
+        }
+      }
+      if (!result.has_value()) {
+        if (state_->finished) {
+          return AsyncGeneratorEnd<T>();
+        }
+        auto new_waiting_fut = Future<T>::Make();
+        state_->waiting_future = new_waiting_fut;
+        guard.Unlock();
+        state_->source().AddCallback(Callback{state_});
+        return new_waiting_fut;
+      }
+      return Future<T>::MakeFinished(*result);
+    }
+  }
+
+ private:
+  struct WrappedComp {
+    bool operator()(const Result<T>& left, const Result<T>& right) {
+      if (!left.ok() || !right.ok()) {
+        return false;
+      }
+      return compare(*left, *right);
+    }
+    Comp compare;
+  };
+
+  struct State {
+    State(AsyncGenerator<T> source, Comp compare, IsNext is_next, T initial_value)
+        : source(std::move(source)),
+          is_next(std::move(is_next)),
+          previous_value(std::move(initial_value)),
+          waiting_future(),
+          queue(WrappedComp{compare}),
+          finished(false),
+          mutex() {}
+
+    AsyncGenerator<T> source;
+    IsNext is_next;
+    T previous_value;
+    Future<T> waiting_future;
+    std::priority_queue<Result<T>, std::vector<Result<T>>, WrappedComp> queue;
+    bool finished;
+    util::Mutex mutex;
+  };
+
+  class Callback {
+   public:
+    explicit Callback(std::shared_ptr<State> state) : state_(std::move(state)) {}
+
+    void operator()(const Result<T> result) {
+      Future<T> to_deliver;
+      bool finished;
+      {
+        auto guard = state_->mutex.Lock();
+        bool ready_to_deliver = false;
+        if (!result.ok()) {
+          while (!state_->queue.empty()) {
+            state_->queue.pop();
```

Comment:
Wow, `std::priority_queue` doesn't allow to clear the queue in a single call? (except by using `std::swap` it seems...)

Reply:
Some googling suggests creating a new queue should work well so I went with that.

Reply:
Nevermind.  Both `state_->queue = {}` and `swap` fail if the callables are lambdas because (I think) of lambda's deleted copy assignment operator.

Reply:
Ah, `= {}` fails because there is no assignment and `swap` fails because I cannot create a queue without an instance of `comp`.  I can create a second empty queue when I create the first one but that is kind of clumsy and it can only swap once (although it should only have to swap once).

At that point, it would probably just be cleaner to have a dedicated spot for the "final result" and call the destructor on the queue instead of clearing it.

I think I'll just keep it how it is.